src/library_browser.js

```diff
@@ -678,23 +694,55 @@ var LibraryBrowser = {
          h = Math.round(h * factor);
       }
       if (Browser.resizeCanvas) {
-        if (canvas.width  != w) canvas.width  = w;
-        if (canvas.height != h) canvas.height = h;
-        if (typeof canvas.style != 'undefined') {
+        wNative = w;
+        hNative = h;
+      }
+      const wNativeScaled = Math.floor(wNative * scale);
+      const hNativeScaled = Math.floor(hNative * scale);
+      if (canvas.width  != wNativeScaled) canvas.width  = wNativeScaled;
+      if (canvas.height != hNativeScaled) canvas.height = hNativeScaled;
+      if (typeof canvas.style != 'undefined') {
+        if (wNativeScaled != wNative || hNativeScaled != hNative) {
+          canvas.style.setProperty( "width", wNative + "px", "important");
+          canvas.style.setProperty("height", hNative + "px", "important");
+        } else {
           canvas.style.removeProperty( "width");
           canvas.style.removeProperty("height");
         }
-      } else {
-        if (canvas.width  != wNative) canvas.width  = wNative;
-        if (canvas.height != hNative) canvas.height = hNative;
-        if (typeof canvas.style != 'undefined') {
-          if (w != wNative || h != hNative) {
-            canvas.style.setProperty( "width", w + "px", "important");
-            canvas.style.setProperty("height", h + "px", "important");
-          } else {
-            canvas.style.removeProperty( "width");
-            canvas.style.removeProperty("height");
-          }
+      }
+    },
+
+    getDevicePixelRatio() {
+      return (typeof devicePixelRatio == 'number' && devicePixelRatio) || 1.0;
+    },
+
+    getHiDPIScale() {
+      return Browser.isHiDPIAware ? Browser.getDevicePixelRatio() : 1.0;
+    },
+
+    devicePixelRatioMQS: null,
+    onDevicePixelRatioChange() {
+      const canvas = Module['canvas'];
+      Browser.updateCanvasDimensions(canvas, canvas.clientWidth, canvas.clientHeight);
+      Browser.updateResizeListeners();
+    },
+
+    setHiDPIAware(isHiDPIAware) {
+      isHiDPIAware = !!isHiDPIAware; // coerce to boolean
+      if (Browser.isHiDPIAware != isHiDPIAware) {
+        Browser.isHiDPIAware = isHiDPIAware;
+        const canvas = Module['canvas'];
+        Browser.updateCanvasDimensions(canvas, canvas.clientWidth, canvas.clientHeight);
+        Browser.updateResizeListeners();
+
+        // handling dynamic changes to devicePixelRatio
+        if (Browser.devicePixelRatioMQS) {
+          Browser.devicePixelRatioMQS.removeEventListener("change", Browser.onDevicePixelRatioChange);
+          Browser.devicePixelRatioMQS = null;
+        }
+        if (Browser.isHiDPIAware) {
+          Browser.devicePixelRatioMQS = window.matchMedia("(resolution: " + Browser.getDevicePixelRatio() + "dppx)");
+          Browser.devicePixelRatioMQS.addEventListener("change", Browser.onDevicePixelRatioChange);
```

Comment:
Use single quotes for strings in this function.   I know we are not consistent everywhere, but we try to use single quotes by default.

Reply:
Done